### PR TITLE
Owasm keeper test  💯

### LINKED ISSUE #1690: Owasm keeper test
There are 2 functions (prepare and execute) write a test to coverage all branch

### PR DESCRIPTION
**Pull request by Benzbeeb**
Fixed #1690

### CODE REVIEW

File: chain/x/oracle/keeper/owasm_test.go [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+	oracleScriptID := k.AddOracleScript(ctx, os)
+	calldata, _ := hex.DecodeString("030000004254436400000000000000")
+	askCount := int64(1)
+	minCount := int64(2)
+	clientID := "beeb"
+	ibcInfo := types.NewIBCInfo("source_port", "source_channel")
+
+	m := types.NewMsgRequestData(oracleScriptID, calldata, askCount, minCount, clientID, Alice.Address)
+	err := k.PrepareRequest(ctx, &m, &ibcInfo)
+	require.NoError(t, err)
```
Comment (sorawit) [inline chain/x/oracle/keeper/owasm_test.go:43]: Check that request exists in the store.
Check that correct events are emitted.

File: chain/x/oracle/keeper/owasm_test.go [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+}
+
+func TestPrepareRequestGetRandomValidatorFail(t *testing.T) {
```
Comment (sorawit) [inline chain/x/oracle/keeper/owasm_test.go]: Validators

File: chain/pkg/owasm/res/beeb.wat [hunk 3]
```diff
@@ -0,0 +1,31 @@
+(module
```
Comment (sorawit) [inline chain/pkg/owasm/res/beeb.wat:1]: A simple test oracle script wasm code that.
- "prepare": Requests external data from sources 1, 2, 3 with call data "beeb"
- "resolve": Resolves with result "beeb"

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/x/oracle/keeper/owasm_test.go [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+
+	require.Equal(t, 4, len(events))
+	require.Equal(t, types.EventTypeRequest, events[0].Type)
+	require.Equal(t, []byte(types.AttributeKeyID), events[0].Attributes[0].Key)
+	require.Equal(t, []byte("1"), events[0].Attributes[0].Value)
+
+	require.Equal(t, []byte(types.AttributeKeyValidator), events[0].Attributes[1].Key)
+	require.Equal(t, []byte(Validator1.ValAddress.String()), events[0].Attributes[1].Value)
+
+	require.Equal(t, types.EventTypeRawRequest, events[1].Type)
```
Comment (taobun) [inline chain/x/oracle/keeper/owasm_test.go]: Move this line to for loop

File: chain/x/oracle/keeper/owasm_test.go [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+
+	oracleScriptID := k.AddOracleScript(ctx, os)
+	calldata, _ := hex.DecodeString("030000004254436400000000000000")
+	askCount := int64(10000)
+	minCount := int64(2)
+	clientID := "beeb"
+	ibcInfo := types.NewIBCInfo("source_port", "source_channel")
+
+	m := types.NewMsgRequestData(oracleScriptID, calldata, askCount, minCount, clientID, Alice.Address)
+	err := k.PrepareRequest(ctx, &m, &ibcInfo)
```
Comment (taobun) [inline chain/x/oracle/keeper/owasm_test.go]: Send nil if this field will be ignored.

File: chain/x/oracle/keeper/owasm_test.go [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+
+	os, clear4 := getTestOracleScript()
+	defer clear4()
+
+	k.AddOracleScript(ctx, os)
+	calldata, _ := hex.DecodeString("030000004254436400000000000000")
+	askCount := int64(1)
+	minCount := int64(2)
+	clientID := "beeb"
+	ibcInfo := types.NewIBCInfo("source_port", "source_channel")
```
Comment (taobun) [inline chain/x/oracle/keeper/owasm_test.go]: ditto

File: chain/x/oracle/keeper/owasm_test.go [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+
+	os, clear4 := getTestOracleScript()
+	defer clear4()
+
+	oracleScriptID := k.AddOracleScript(ctx, os)
+	calldata, _ := hex.DecodeString("030000004254436400000000000000")
+	askCount := int64(1)
+	minCount := int64(2)
+	clientID := "beeb"
+	ibcInfo := types.NewIBCInfo("source_port", "source_channel")
```
Comment (taobun) [inline chain/x/oracle/keeper/owasm_test.go]: ditto

**Review by taobun — CHANGES_REQUESTED** [review]

File: chain/x/oracle/keeper/owasm_test.go [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+
+	oracleScriptID := k.AddOracleScript(ctx, os)
+	calldata, _ := hex.DecodeString("030000004254436400000000000000")
+	askCount := uint64(1)
+	minCount := uint64(2)
+	clientID := "beeb"
+	ibcInfo := types.NewIBCInfo("source_port", "source_channel")
+
+	m := types.NewMsgRequestData(oracleScriptID, calldata, askCount, minCount, clientID, Alice.Address)
+	err := k.PrepareRequest(ctx, &m, &ibcInfo)
```
Comment (taobun) [inline chain/x/oracle/keeper/owasm_test.go]: nil

File: chain/x/oracle/keeper/owasm_test.go [hunk 9]
```diff
@@ -0,0 +1,300 @@
...
+		require.Equal(t, []byte(exIDs[idx]), events[idx].Attributes[2].Value)
+		require.Equal(t, []byte(types.AttributeKeyCalldata), events[idx].Attributes[3].Key)
+		require.Equal(t, []byte("beeb"), events[idx].Attributes[3].Value)
+	}
+}
+
+func TestPrepareRequestInvalidAskCountFail(t *testing.T) {
+	_, ctx, k := createTestInput()
+	ctx = ctx.WithBlockTime(time.Unix(int64(1581589790), 0))
+
```
Comment (taobun) [inline chain/x/oracle/keeper/owasm_test.go:85]: Set params here

File: chain/x/oracle/keeper/owasm_test.go [hunk 10]
```diff
@@ -0,0 +1,300 @@
...
+	askCount := uint64(1)
+	minCount := uint64(2)
+	clientID := "beeb"
+
+	m := types.NewMsgRequestData(oracleScriptID, calldata, askCount, minCount, clientID, Alice.Address)
+	err := k.PrepareRequest(ctx, &m, nil)
+	require.Error(t, err)
+}
+
+func TestPrepareRequestGetDataSourceFail(t *testing.T) {
```
Comment (taobun) [inline chain/x/oracle/keeper/owasm_test.go:268]: What different with TestPrepareRequestBadWasmExecutionFail

**Review by taobun — CHANGES_REQUESTED** [review]

**Review by sorawit — APPROVED** [review]

**Review by taobun — APPROVED** [review]

**Pull request merged**